README.md
# UUID Generator

This extension generates Universal Unique Identifier (UUID) values and will either insert them into the current active editor, or place them on the clipboard.

One useful case for this extension is to create UUIDs for configuring new profiles for [Windows Terminal](https://docs.microsoft.com/en-us/windows/terminal/).

## Features

This extension supports creating a new version 4 UUID and then either adding it to the current location in the active editor, or placing it on the clipboard.

The _Insert new UUID_ function is available from the editor context menu, or as a keyboard binding. By default, the keyboard binding is `Alt+Shift+U`.

The other way to run the commands is from the command palette by pressing (`Ctrl+Shift+P` or `Cmd+Shift+P` on Mac), typing `UUID` and selecting the desired function.
* Copy new UUID to the clipboard
* Insert new UUID
* Insert a new 'nil' UUID (where all the digits are 0)

Through [configuration settings](#settings), it is possible to wrap the created UUID with braces or quotes when it is created, for example:
* `b26e67ee-0a43-4c3f-8aea-a1179427faf7`
* `'b26e67ee-0a43-4c3f-8aea-a1179427faf7'`
* `"b26e67ee-0a43-4c3f-8aea-a1179427faf7"`
* `{b26e67ee-0a43-4c3f-8aea-a1179427faf7}`
* `[b26e67ee-0a43-4c3f-8aea-a1179427faf7]`
* `(b26e67ee-0a43-4c3f-8aea-a1179427faf7)`

This extension works as expected when inserting a new UUID into the active text editor. It will insert the UUID at the location of the caret in the current active editor, replace any existing text selection if there is one, and generate multiple UUIDs and insert them all when a [multi-cursor selection](https://code.visualstudio.com/docs/getstarted/tips-and-tricks#_multi-cursor-selection) is active.

### UUID type
The generated UUIDs are type 4, variation 1, which means unique based on random numbers. The extension can also generate a nil UUID, where all the digits are 0.

See the [release notes](#release-notes) for more information.

## Requirements

* Requires VS Code v1.50+. This will be reviewed

## Settings

| Name | Description | Default |
|--|--|--|
| `vscode-uuid-generator.contextMenu.enabled` | Whether the _Insert new UUID_ function is available from the editor context menu | `true` |
| `vscode-uuid-generator.uppercaseDigits` | Whether the hex digits are forced to uppercase (`true`) or lowercase (`false`) | `false` |
| `vscode-uuid-generator.textSelection`   | When the UUID is inserted into the text, this setting controls whether the pasted value is marked as selected (`true`) or whether the selection is cleared (`false`) | `false` |
| `vscode-uuid-generator.decorationStyle` | Whether the generated UUID should be automatically wrapped with quotes or braces. Possible values:<ul><li>`none`</li><li>`surroundSingleQuotes`</li><li>`surroundDoubleQuotes`</li><li> `surroundCurlyBraces`</li><li>`surroundRoundedBraces`</li><li>`surroundSquareBraces`</ul> | `none` |

## Known Limitations and Issues

TBD

## Release Notes

This extension generates UUIDs that are version 4, variation 1 according to the specification [RFC4122 section 4.4](https://tools.ietf.org/html/rfc4122#section-4.4). Version 4 UUIDs are generated with random numbers as generated by VS Code and modified by a portion of the current timestamp to add another random element. This cannot guarantee uniqueness to the same degree as something tied to hardware identifiers such as MAC address, but it is _statistically unlikely_ that any two UUIDs would be identical.

This extension can also be used to produce a 'nil' UUID, one where all the values are zero. See [RFC4122 section 4.1.7](https://tools.ietf.org/html/rfc4122#section-4.1.7)

### References
* [RFC4122 - A Universally Unique IDentifier (UUID) URN Namespace](https://www.ietf.org/rfc/rfc4122.txt)
* [Wikipedia - Universally unique identifier](https://en.wikipedia.org/wiki/Universally_unique_identifier)


CHANGELOG.md
# Change Log

All notable changes to the "vscode-theavonian-uuid-generator" and "vscode-uuid-generator" extension will be documented in this file.

## [0.4.2] - 2023-09-23
Implemented support for VS Code on the web.

## [0.4.1] - 2022-03-28
Implemented support for VS Code on the web.

## [0.4.0] - 2022-03-26
Updated baseline component dependencies to prepare for VS Code web implementation

## [0.3.0] - 2021-09-16
Declared support for virtual workspaces and untrusted workspaces
Fixed year portion of release dates in this change log.

## [0.2.1] - 2020-11-08
Fixed change log and updated readme for the 0.2 release.

## [0.2.0] - 2020-11-08
### Added
* Editor context menu now includes the _Insert new UUID_ function. This can be controlled through the configuration of the extension.
* Keybinding added. Alt+Shift+U will invoke _Insert new UUID_ into the active editor. This can be remapped in the keyboard shortcuts editor.

## [0.1.1] - 2020-10-18
Minor corrections to documentation and extension setup. No changes to functionality.

## [0.1.0] - 2020-10-17
Some fundamental improvements to the usabilty of the component through configuration settings.
### Added
- Configuration items for UUID formatting - lower/uppercase and whether the UUID value should be surrounded by quotes or braces
- Configuraiton item for whether the UUID value, when inserted into the active editor, should be marked as selected
- Added a function to insert a nil UUID (where all the digits are zero)

## [0.0.1] - 2020-10-16
Initial release to go through the process of preparing and uploading
### Added
- Minimal viable functionality
    - Copy UUID to clipboard
    - Insert UUID into current editor

[0.4.1]: https://github.com/Motivesoft/vscode-uuid-generator/releases/tag/v0.4.1
[0.4.0]: https://github.com/Motivesoft/vscode-uuid-generator/releases/tag/v0.4.0
[0.3.0]: https://github.com/Motivesoft/vscode-uuid-generator/releases/tag/v0.3.0
[0.2.1]: https://github.com/Motivesoft/vscode-uuid-generator/releases/tag/v0.2.1
[0.2.0]: https://github.com/Motivesoft/vscode-uuid-generator/releases/tag/v0.2.0
[0.1.1]: https://github.com/Motivesoft/vscode-uuid-generator/releases/tag/v0.1.1
[0.1.0]: https://github.com/Motivesoft/vscode-uuid-generator/releases/tag/v0.1.0
[0.0.1]: https://github.com/Motivesoft/vscode-uuid-generator/releases/tag/v0.0.1


webpack.config.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

//@ts-check
'use strict';

//@ts-check
/** @typedef {import('webpack').Configuration} WebpackConfig **/

const path = require('path');
const webpack = require('webpack');

/** @type WebpackConfig */
const webExtensionConfig = {
	mode: 'production', // this leaves the source code as close as possible to the original (when packaging we set this to 'production')
	target: 'webworker', // extensions run in a webworker context
	entry: {
		'extension': './src/web/extension.ts',
		'test/suite/index': './src/web/test/suite/index.ts'
	},
	output: {
		filename: '[name].js',
		path: path.join(__dirname, './out/web'),
		libraryTarget: 'commonjs',
		devtoolModuleFilenameTemplate: '../../[resource-path]'
	},
	resolve: {
		mainFields: ['browser', 'module', 'main'], // look for `browser` entry point in imported node modules
		extensions: ['.ts', '.js'], // support ts-files and js-files
		alias: {
			// provides alternate implementation for node module and source files
		},
		fallback: {
			// Webpack 5 no longer polyfills Node.js core modules automatically.
			// see https://webpack.js.org/configuration/resolve/#resolvefallback
			// for the list of Node.js core module polyfills.
			'assert': require.resolve('assert')
		}
	},
	module: {
		rules: [{
			test: /\.ts$/,
			exclude: /node_modules/,
			use: [{
				loader: 'ts-loader'
			}]
		}]
	},
	plugins: [
		new webpack.ProvidePlugin({
			process: 'process/browser', // provide a shim for the global `process` variable
		}),
	],
	externals: {
		'vscode': 'commonjs vscode', // ignored because it doesn't exist
	},
	performance: {
		hints: false
	},
	devtool: 'nosources-source-map', // create a source map that points to the original source file
	infrastructureLogging: {
		level: "log", // enables logging required for problem matchers
	},
};

module.exports = [ webExtensionConfig ];

src/common.ts
import * as vscode from 'vscode';

export function configureContext(context: vscode.ExtensionContext) {
    // Configure the editor context
	console.log('Configuring context for "vscode-theavonian-uuid-generator"');

    // See https://en.wikipedia.org/wiki/Universally_unique_identifier
    // See https://www.ietf.org/rfc/rfc4122.txt, section 4.4

	let insertDefault = vscode.commands.registerCommand('vscode-theavonian-uuid-generator.insertDefault', () => {
		// Generate the UUID and insert it at the current edit point
		let editor = vscode.window.activeTextEditor;
        if (editor) {
			editor.edit( edit => {
				editor?.selections.forEach( v => edit.replace( v, formatGuid( makeGuid() ) ) );
			} ).then( success => {
				var select = vscode.workspace.getConfiguration().get("vscode-theavonian-uuid-generator.textSelection");
				if( success && editor && !select ) {
					editor.selection = new vscode.Selection( editor.selection.end, editor.selection.end );
				}
			} );
		}
	});

    let insertHandle = vscode.commands.registerCommand('vscode-theavonian-uuid-generator.insertHandle', () => {
		// Generate the UUID and insert it at the current edit point
		let editor = vscode.window.activeTextEditor;
        if (editor) {
			editor.edit( edit => {
				editor?.selections.forEach( v => edit.replace( v, makeHandle() ) );
			} ).then( success => {
				var select = vscode.workspace.getConfiguration().get("vscode-theavonian-uuid-generator.textSelection");
				if( success && editor && !select ) {
					editor.selection = new vscode.Selection( editor.selection.end, editor.selection.end );
				}
			} );
		}
	});
	
	let insertNil = vscode.commands.registerCommand('vscode-theavonian-uuid-generator.insertNil', () => {
		// Generate the UUID and insert it at the current edit point
		let editor = vscode.window.activeTextEditor;
        if (editor) {
			editor.edit( edit => {
				editor?.selections.forEach( v => edit.replace( v, formatGuid( makeNilGuid() ) ) );
			} ).then( success => {
				var select = vscode.workspace.getConfiguration().get("vscode-theavonian-uuid-generator.textSelection");
				if( success && editor && !select ) {
					editor.selection = new vscode.Selection( editor.selection.end, editor.selection.end );
				}
			} );
		}
	});
	
	let copyDefault = vscode.commands.registerCommand('vscode-theavonian-uuid-generator.copyDefault', () => {
		// Generate UUID and add it to the clipboard
		vscode.env.clipboard.writeText( formatGuid( makeGuid() ) );
		
		// Display a message box to the user
		vscode.window.showInformationMessage('UUID copied to clipboard');
	});

    let copyHandle = vscode.commands.registerCommand('vscode-theavonian-uuid-generator.copyHandle', () => {
		// Generate UUID and add it to the clipboard
		vscode.env.clipboard.writeText( makeHandle() );
		
		// Display a message box to the user
		vscode.window.showInformationMessage('Handle copied to clipboard');
	});

	context.subscriptions.push(insertDefault);
	context.subscriptions.push(insertHandle);
	context.subscriptions.push(insertNil);
	context.subscriptions.push(copyDefault);
	context.subscriptions.push(copyHandle);
}

export function deconfigureContext() {
    // Nothing to do
}

function formatGuid(uuid: string) {
	var result: string;

	var decorationStyle = vscode.workspace.getConfiguration().get("vscode-theavonian-uuid-generator.decorationStyle");

	switch( decorationStyle ) {
		case "surroundSingleQuotes":
			result = `'${uuid}'`;
			break;
			
		case "surroundDoubleQuotes":
			result = `\"${uuid}\"`;
			break;
		
		case "surroundCurlyBraces":
			result = `{${uuid}}`;
			break;
			
		case "surroundRoundedBraces":
			result = `(${uuid})`;
			break;
			
		case "surroundSquareBraces":
			result = `[${uuid}`;
			break;

		default:
		case "none":
			result = uuid;
			break;
	}

	return result;
}

function makeNilGuid() {
    // See https://www.ietf.org/rfc/rfc4122.txt, section 4.1.7
	return "00000000-0000-0000-0000-000000000000";
}

function makeHandle() {
    let h = makeGuid();
	return `h${h.replace(RegExp("-", 'g'), 'g')}`;
}

function makeGuid() {
	var result: string;
	var i: string;
	var j: number;
	var n: number;
	
    // 32-digit hex number in the style:
    // 	xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx
    // Where:
    //  M shall be 4 to indicate this as a Version 4 (random) UUID
    //  N shall be 10 (binary) to indicate this as a Variant 1
    
    // See https://www.ietf.org/rfc/rfc4122.txt, section 4.4

    // We're going to use a random number generator, but then introduce a time
    // influence to make sure we don't get caught up in some duplication issue
    // related to the pseudo-random number generator
	var dt = new Date().getTime();

	// Build this up over the course of this method
	result = "";

	// Generate each digit individually
	for (j = 0; j < 32; j++) {
		// Create the next 4 bit number for the overall string.
		// 4 bits = decimal 0-15, hex digit 0-F, binary 0000-1111

		if( j === 12 ) {
			// Set 'M' to version 4 (0100 binary)
			// This occupies all 4 bits, so we replace the whole value
			n = 0x04;
		}
		else {
            // Work out the next hex digit (random number + some ad-hoc part of a timestamp modulo 16)
            n = Math.floor(dt + Math.random() * 16) % 16;

            // Work through the timestamp value by modifying it until it is exhausted
            dt = Math.floor(dt/16);
			
			if( j === 16 ) {
				// Set 'N' to variant 1 (10xx binary)
				// This occupies the top 2 bits only, so mask and replace them 
				n &= 0x03;
				n |= 0x08;
			}
		}

		// Format the string with hyphens for legibility
		if (j === 8 || j === 12 || j === 16 || j === 20) {
			result = result + '-';
		}

		// Add the hex digit to the result
		i = n.toString(16);
		result = result + i;
	}

	// Return the result as upper/lowercase as configured
	var uppercase = vscode.workspace.getConfiguration().get("vscode-theavonian-uuid-generator.uppercaseDigits");

	return uppercase ? result.toUpperCase() : result.toLowerCase();
}
